### ISSUE TITLE
`PossiblyUndefinedIntArrayOffset` is not reported in the class method with no key checking, but works outside

### ISSUE THREAD

**Issue by AlexMinaev19** [vimeo/psalm#9609]
Imagine we have a container for some elements that are stored by key
```php
class B {}

class A {
    /**
     * @param array<non-empty-string, B> $container
     */
    public function __construct(
        public array $container
    ) {
    }

    ...
}
```
And we want to do typical operations: `add`, `get`, `has`.

```php
...

    /**
     * @param non-empty-string $k
     */
    public function add(string $k, B $b): void
    {
        $this->container[$k] = $b;
    }
    
    /**
     *
     * @param non-empty-string $k
     */
    public function has(string $k): bool
    {
        return array_key_exists($k, $this->container);
    }
    
    /**
     * @param non-empty-string $k
     *
     * @return B
     */
    public function get(string $k): B
    {
        return $this->container[$k];
    }

...
```

Here we see a potential problem in the method `A::get`: we do not check that `container` has the provided key.
So, potentially it can throw an exception and this code is dangerous.
Unfortunately, `psalm` does not see a problem here. But if you will do the same outside of the class it shows an error.

https://psalm.dev/r/522a167d14

**Comment by lxix** [comment]
@ygottschalk I'm not completely sure about that. Take a look at this one: https://psalm.dev/r/c0275b936d

**Comment by orklah** [comment]
Yeah, @ygottschalk got the right answer. Psalm will check that a specific literal key exists in the array shape. If you have a generic `string` type, there's nothing to check for Psalm
